lib/opFns.js

```diff
@@ -1024,3 +1015,73 @@ function makeCall (runState, callOptions, localOpts, cb) {
+  if (runState._common.gteHardfork('constantinople')) {
+    async.parallel({
+      original: function (cb) { runState.originalState.getContractStorage(address, key, cb) },
+      current: function (cb) { runState.stateManager.getContractStorage(address, key, cb) }
+    }, cb)
+  } else {
+    runState.stateManager.getContractStorage(address, key, cb)
+  }
+}
+
+function updateSstoreGas (runState, found, value) {
+  if (runState._common.gteHardfork('constantinople')) {
+    var original = found.original
+    var current = found.current
+    if (current.equals(value)) {
+      // If current value equals new value (this is a no-op), 200 gas is deducted.
+      subGas(runState, new BN(runState._common.param('gasPrices', 'netSstoreNoopGas')))
+      return
+    }
+    // If current value does not equal new value
+    if (original.equals(current)) {
+      // If original value equals current value (this storage slot has not been changed by the current execution context)
+      if (original.length === 0) {
+        // If original value is 0, 20000 gas is deducted.
+        return subGas(runState, new BN(runState._common.param('gasPrices', 'netSstoreInitGas')))
+      }
+      if (value.length === 0) {
+        // If new value is 0, add 15000 gas to refund counter.
+        runState.gasRefund = runState.gasRefund.addn(runState._common.param('gasPrices', 'netSstoreClearRefund'))
+      }
+      // Otherwise, 5000 gas is deducted.
+      return subGas(runState, new BN(runState._common.param('gasPrices', 'netSstoreCleanGas')))
+    }
+    // If original value does not equal current value (this storage slot is dirty), 200 gas is deducted. Apply both of the following clauses.
+    if (original.length !== 0) {
+      // If original value is not 0
+      if (current.length === 0) {
+        // If current value is 0 (also means that new value is not 0), remove 15000 gas from refund counter. We can prove that refund counter will never go below 0.
+        runState.gasRefund = runState.gasRefund.subn(runState._common.param('gasPrices', 'netSstoreClearRefund'))
+      } else if (value.length === 0) {
+        // If new value is 0 (also means that current value is not 0), add 15000 gas to refund counter.
+        runState.gasRefund = runState.gasRefund.addn(runState._common.param('gasPrices', 'netSstoreClearRefund'))
+      }
+    }
+    if (original.equals(value)) {
+      // If original value equals new value (this storage slot is reset)
+      if (original.length === 0) {
+        // If original value is 0, add 19800 gas to refund counter.
+        runState.gasRefund = runState.gasRefund.addn(runState._common.param('gasPrices', 'netSstoreResetClearRefund'))
+      } else {
+        // Otherwise, add 4800 gas to refund counter.
+        runState.gasRefund = runState.gasRefund.addn(runState._common.param('gasPrices', 'netSstoreResetRefund'))
+      }
+    }
+    return subGas(runState, new BN(runState._common.param('gasPrices', 'netSstoreDirtyGas')))
```

Comment:
This is working down the various cases from the EIP and adding to the existing ``runState.gasRefund`` variable. I assume this is correct since tests are passing here on a broad basis.

If there are smaller caveats or edge cases we can do subsequent PRs to correct once we have a clearer picture on the test runs and tests have stabilized more.